// typed-event-target/dist/generic-listen-target.js
/**
 * A generic listen target that stores listeners keyed by event constructor rather than event type
 * string. Listeners are matched via `instanceof`, so class hierarchy is respected: a listener
 * registered for a parent class will also fire for subclass instances.
 *
 * Unlike `TypedListenTarget`, this class accepts any `Event` without a constraining type parameter.
 *
 * @category Main
 */
export class GenericListenTarget {
    listeners = new Map();
    /**
     * Get a count of all currently attached listeners. If a listener is removed, it will no longer
     * be counted.
     */
    getListenerCount() {
        const maps = [...this.listeners.values()];
        return maps.reduce((accum, innerMap) => {
            return accum + innerMap.size;
        }, 0);
    }
    /**
     * Listen to events that are instances of the given event class. Because matching uses
     * `instanceof`, a listener for a parent class will also fire for subclass instances.
     *
     * @returns A callback to remove the listener.
     */
    listen(eventClass, listenerCallback, options = {}) {
        const genericCallback = listenerCallback;
        const innerMap = this.getOrCreateInnerMap(eventClass);
        const removeListener = () => {
            return innerMap.delete(genericCallback);
        };
        function wrappedCallback(event, removeSelf) {
            if (options.once) {
                removeListener();
            }
            genericCallback(event, removeSelf);
        }
        innerMap.set(genericCallback, {
            listener: wrappedCallback,
            removeListener,
        });
        return removeListener;
    }
    /**
     * Remove a previously attached listener.
     *
     * @returns Whether the listener existed and was removed.
     */
    removeListener(eventClass, listenerCallback) {
        const genericCallback = listenerCallback;
        const innerMap = this.listeners.get(eventClass);
        if (!innerMap) {
            return false;
        }
        const wrapper = innerMap.get(genericCallback);
        if (!wrapper) {
            return false;
        }
        return wrapper.removeListener();
    }
    /**
     * Dispatch an event. Fires all listeners whose registered event class matches the dispatched
     * event via `instanceof`.
     *
     * @returns The number of listeners that were fired.
     */
    dispatch(event) {
        if (event.target == undefined) {
            Object.defineProperty(event, 'target', {
                writable: false,
                value: this,
            });
        }
        let firedCount = 0;
        this.listeners.forEach((innerMap, eventClass) => {
            if (event instanceof eventClass) {
                /**
                 * Capture size before invoking listeners since listeners may remove themselves
                 * during iteration.
                 */
                firedCount += innerMap.size;
                innerMap.forEach((wrapper) => {
                    wrapper.listener(event, wrapper.removeListener);
                });
            }
        });
        return firedCount;
    }
    /**
     * Remove all currently attached listeners.
     *
     * @returns The number of listeners that were removed.
     */
    removeAllListeners() {
        const maps = [...this.listeners.values()];
        const totalRemoved = maps.reduce((accum, innerMap) => {
            const size = innerMap.size;
            innerMap.clear();
            return accum + size;
        }, 0);
        this.listeners.clear();
        return totalRemoved;
    }
    /** Remove all internal state to free up resources. */
    destroy() {
        this.removeAllListeners();
    }
    /** Retrieve the inner listener map for the given event class, creating it if it doesn't exist. */
    getOrCreateInnerMap(eventClass) {
        const existing = this.listeners.get(eventClass);
        if (existing) {
            return existing;
        }
        const innerMap = new Map();
        this.listeners.set(eventClass, innerMap);
        return innerMap;
    }
}
